### PR TITLE
Add actively generated sygus enumerators

### PR DESCRIPTION
**Pull request by ajreynol**
Implements a new paradigm for "actively-generated enumerators" in the sygus solver, and adds option --sygus-var-agnostic that enables (wherever applicable) variable-agnostic enumerators, which are an instance of actively-generated enumerators. This option is disabled by default for now.

### CODE REVIEW

**Comment by HanielB** [comment]
Add the data on to which grammars / depths we have checked that the enumerations produce the same number of unique terms.

**Comment by ajreynol** [comment]
Variable-agnostic enumerators are expected to be a factor N! faster than default enumerators for each group of variables of size N with the same subclass. This performance has been roughly confirmed, for example:

37 seconds ---> 25 seconds, for 2 variable string-term grammar depth 2
162 seconds ---> 32 seconds, for 3 variable string-term grammar depth 2

--sygus-stream has confirmed that the number of solutions with and without --sygus-var-agnostic for fixed depth on grammars remains unchanged:

string-term, depth up to 2
bv-term, depth up to 3

This indicates that the new enumeration is "sound", i.e. it does not overly prune the search space.

File: src/theory/quantifiers/sygus/synth_conjecture.h [hunk 1]
```diff
@@ -174,6 +182,46 @@ class SynthConjecture
...
+  bool getEnumeratedValues(std::vector<Node>& n, std::vector<Node>& v);
+  /**
+   * Get model value for term n. If n has a value that was excluded by
+   * datatypes sygus symmetry breaking, this method returns null.
+   */
+  Node getEnumeratedValue(Node n);
+  /** enumerator generators for each actively-generated enumerator */
+  std::map<Node, std::unique_ptr<EnumValGenerator> > d_evg;
+  /**
+   * Map from enumerators to whether they are currently being
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_conjecture.h]: clarify that the map is from enum to the abstract value you are currently enumerating concrete values for, and that when it's null you're no longer in "actively-generating" state

File: src/theory/quantifiers/sygus/synth_conjecture.h [hunk 2]
```diff
@@ -174,6 +182,46 @@ class SynthConjecture
...
+   * "actively-generated". That is, we are in a state where we have called
+   * d_evg[e].addValue(v) for some v, and d_evg[e].getNext() has not yet
+   * returned null.
+   */
+  std::map<Node, Node> d_ev_curr_active_gen;
+  /** the current waiting value of each actively-generated enumerator, if any
+   *
+   * This caches values are actively generated and that we have not yet
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_conjecture.h]: that are?

File: src/theory/quantifiers/sygus/synth_conjecture.cpp [hunk 3]
```diff
@@ -646,13 +681,101 @@ Node SynthConjecture::getEnumeratedValue(Node e)
...
+    else
+    {
+      Assert(false);
+    }
+    Node lem = exp.size() == 1 ? exp[0] : nm->mkNode(OR, exp);
+    Trace("cegqi-lemma") << "Cegqi::Lemma : actively-generated enumerator "
+                            "exclude current solution : "
+                         << lem << std::endl;
+    Trace("sygus-active-gen-debug")
+        << "Active-gen: block " << absE << std::endl;
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_conjecture.cpp]: use the `TermDbSygus::toStreamSygus`

**Comment by ajreynol** [comment]
Notice that this PR also has addressed several existing issues with combining --decision=internal with sygus. It also fixes symmetry breaking for variable-agnostic enumerators in a few ways that came up while testing this branch.

File: src/theory/datatypes/datatypes_sygus.cpp [hunk 4]
```diff
@@ -1127,11 +1149,22 @@ Node SygusSymBreakNew::registerSearchValue(
...
         unsigned prev_sz = d_cache[a].d_search_val_sz[tn][bad_val_bvr];
-        if( prev_sz>sz ){
+        bool doFlip = (prev_sz > sz);
+        if (doFlip)
+        {
           //swap : the excluded value is the previous
           d_cache[a].d_search_val_sz[tn][bad_val_bvr] = sz;
           bad_val = d_cache[a].d_search_val[tn][bad_val_bvr];
           bad_val_o = nv;
+          if (Trace.isOn("sygus-sb-exc"))
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_sygus.cpp]: you do not need to guard this with `Trace.isOn`. The call to `quantifiers::TermDbSygus::toStreamSygus` already does this test and the other trace instructions will as well.

Comment (ajreynol) [reply src/theory/datatypes/datatypes_sygus.cpp]: This still saves two calls and 2 checks.

**Comment by ajreynol** [comment]
Current status for sygus comp 2018 benchmarks:
```
                           Configuration  #unsat      time    #sat      time #solved    #total
         CVC4-100218-2-ae_sygus_d_status    1201   66148.6       0         0      1201      1490
    CVC4-100218-2-ae_sygus_nrc_di_status    1229   61376.4       0         0      1229      1490
        CVC4-100218-2-ae_sygus_di_status    1211     59279       0         0      1211      1490
CVC4-100218-2-ae_sygus_sva_nrc_di_status    1224   61829.7       0         0      1224      1490
   CVC4-100218-2-ae_sygus_sva_nrc_status    1192   63037.3       0         0      1192      1490
    CVC4-100218-2-ae_sygus_sva_di_status    1204   61787.6       0         0      1204      1490
       CVC4-100218-2-ae_sygus_sva_status    1183   60638.9       0         0      1183      1490
       CVC4-100218-2-ae_sygus_nrc_status    1204   63554.5       0         0      1204      1490
```

**Comment by ajreynol** [comment]
For lustre-sygus:
```
 Configuration,#unsat,time,#sat,time,#solved,#total
         d,185,5823.27,0,0,185,485
    nrc_di,250,49198.2,0,0,250,485
        di,158,3924.75,0,0,158,485
sva_nrc_di,240,12436,0,0,240,485
   sva_nrc,225,20233.5,0,0,225,485
    sva_di,165,2241.16,0,0,165,485
       sva,190,7696.56,0,0,190,485
       nrc,227,24892.8,0,0,227,485
```

**Pull request merged**